ion-schema/src/model/constraints/valid_values.rs

```diff
@@ -0,0 +1,363 @@
+        ValidValuesArgument::Value(value.into())
+    }
+}
+impl From<IonSchemaRange<Decimal>> for ValidValuesArgument {
+    fn from(value: IonSchemaRange<Decimal>) -> Self {
+        ValidValuesArgument::NumericRange(value)
+    }
+}
+impl From<IonSchemaRange<Timestamp>> for ValidValuesArgument {
+    fn from(value: IonSchemaRange<Timestamp>) -> Self {
+        ValidValuesArgument::TimestampRange(value)
+    }
+}
+
+/// Helper type to allow reading any type of numeric value for valid values number ranges.
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+struct NumberRangeValue(Decimal);
+
+impl<V: IslVersion> ReadFromIsl<V> for NumberRangeValue {
+    fn try_read(ion: &Element, ctx: &LoaderContext<V>) -> IonSchemaResult<Self> {
+        let value = match ion.value() {
+            Value::Int(i) => NumberRangeValue(Decimal::from(*i)),
+            Value::Float(f) if f.is_finite() => NumberRangeValue(Decimal::try_from(*f)?),
+            Value::Decimal(d) => NumberRangeValue(*d),
+            other => invalid_schema_error(format!("Not a valid number range boundary: {other}"))?,
+        };
+        Ok(value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::model::constraints::valid_values::ValidValues;
+    use crate::model::{IonSchemaRange, TypeDefinitionBuilder};
+    use crate::test_harness::*;
+    use crate::{ISL_1_0, ISL_2_0};
+    use ion_rs::{Decimal, IonType, Value};
+
+    #[test]
+    fn test_builder_valid_values() {
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .valid_values([Value::Float(1.0), Value::String("foo".into())])
+            .build();
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![ValidValues {
+                values: vec![Value::Float(1.0).into(), Value::String("foo".into()).into(),]
+            }
+            .into()]
+        );
+
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .valid_values(["foo", "bar"])
+            .build();
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![ValidValues {
+                values: vec![
+                    Value::String("foo".into()).into(),
+                    Value::String("bar".into()).into(),
+                ]
+            }
+            .into()]
+        );
+    }
+
+    #[test]
+    fn test_builder_valid_number_range() {
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .valid_number_range(1..=10)
+            .build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![ValidValues {
+                values: vec![IonSchemaRange::from(Decimal::from(1)..=Decimal::from(10)).into()]
+            }
+            .into()]
+        );
+    }
+
+    test_harness!(write_as_isl:
+        ISL_1_0, ISL_2_0 {
+            #[case::empty(
+                Ok("[]"),
+                ValidValues { values: vec![] }
+            )]
```

Comment:
🗺️ Here is a test case for the test harness. I've modeled it after the `#[case]` syntax from `rstest` so that it looks familiar, but if you think it could be easier to read with some other syntax, I can easily change it.

I also considered `test_case_name => (arg_1, arg_2),` for the test cases, but I don't have a particular preference either way.